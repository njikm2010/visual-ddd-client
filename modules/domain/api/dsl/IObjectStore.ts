import { Entity, Node, ValueObject } from './domain-model';
import { DTO } from './query-model';
import { DataObject } from './data-model';

export interface IObjectStore {
  getObjectById: (id: string) => { object: Entity | ValueObject | DTO | DataObject; parent?: Node };
}
