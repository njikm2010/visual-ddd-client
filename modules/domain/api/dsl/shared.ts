import { NameDSL } from '@/modules/domain/domain-design/dsl/dsl';
import { MapTypeRoot } from '@/lib/editor/Model/NodeYMap';

export interface CoreProperties {
  __node_name__: string;
  uuid?: string;
}

export interface Tree<T extends CoreProperties = CoreProperties> {
  id: string;
  parent?: string;
  children: Record<string, never>;
  properties: T;
}

export const ROOT = MapTypeRoot;

export abstract class BaseContainer {
  private postTraverses: (() => void)[] = [];

  constructor() {}

  abstract handle(node: Tree, tree: Record<string, Tree>): void;

  protected addPostTraverse(fn: () => void) {
    this.postTraverses.push(fn);
  }

  traverse(tree: Record<string, Tree>) {
    const walk = (node: Tree) => {
      this.handle(node, tree);

      const childrenKeys = Object.keys(node.children);
      if (childrenKeys) {
        for (const childKey of childrenKeys) {
          node = tree[childKey];
          if (node) {
            walk(node);
          }
        }
      }
    };

    const root = tree[ROOT];
    walk(root);

    if (this.postTraverses.length) {
      const list = this.postTraverses;
      this.postTraverses = [];
      for (const fn of list) {
        fn();
      }
    }
  }
}

export interface IContainer {
  getNodeById(id: string): Node<any> | undefined;
}

/**
 * 基础节点信息
 */
export abstract class Node<T extends NameDSL = NameDSL> {
  /**
   * 唯一 id
   */
  id: string;

  /**
   * 属性
   */
  properties: T;

  container: IContainer;

  get name() {
    return this.properties.name;
  }

  constructor(id: string, properties: T, container: IContainer) {
    this.container = container;
    this.id = id;

    this.properties = properties;
  }

  getReference = (id: string): NameDSL => {
    return this.container.getNodeById(id)!.properties;
  };
}
