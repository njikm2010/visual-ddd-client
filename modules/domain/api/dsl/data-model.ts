/**
 * 数据模型 DSL 转换
 */
import * as DSL from './interface';
import * as ViewDSL from '@/modules/domain/data-design/dsl/dsl';
import { DataObjectName } from '@/modules/domain/data-design/dsl/constants';

import { transformMeta, transformString } from './domain-model';
import { Tree, BaseContainer, IContainer, Node } from './shared';

const DataObjectTypeName = ViewDSL.DataObjectTypeName;

export function transformType(
  type: ViewDSL.DataObjectTypeDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): Record<string, any> {
  const typeName = type.type;

  switch (typeName) {
    case DataObjectTypeName.Date:
    case DataObjectTypeName.DateTime:
    case DataObjectTypeName.Timestamp:
      return {
        ...type,
        defaultValue: transformString(type.defaultValue),
      };
    case DataObjectTypeName.Reference:
      const refProperty = getReference(type.target!, type.targetProperty!);
      return {
        type: refProperty.type.type,
      };
    default:
      return type;
  }
}

export function transformProperty(
  property: ViewDSL.DataObjectPropertyDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataPropertyDSL {
  const { uuid, title, name, description, meta, propertyName, primaryKey, notNull, type } = property;

  return {
    uuid,
    name,
    title,
    description,
    meta: transformMeta(meta),
    propertyName: transformString(propertyName),
    ...(transformType(type, getReference) as {
      type: DSL.DataPropertyType;
      defaultValue?: any;
      autoIncrement?: boolean;
      precision?: number;
      scale?: number;
      length?: number;
    }),
    notNull: primaryKey ? true : notNull,
  };
}

export function transformIndex(
  index: ViewDSL.DataObjectIndexDSL,
  getProperty: (id: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataIndexDSL {
  const { uuid, title, name, description, meta, type, method, properties } = index;

  return {
    uuid,
    name,
    title,
    description,
    meta: transformMeta(meta),
    type,
    method,
    properties: properties.map(i => getProperty(i).name),
  };
}

export function transformDataObjectReference(
  viewDSL: ViewDSL.DataObjectDSL,
  getObject: (objectId: string) => ViewDSL.DataObjectDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataObjectReferenceDSL {
  const properties = viewDSL.properties;
  const targets: Map<
    // 目标对象 name
    string,
    {
      cardinality: 'OneToOne' | 'OneToMany' | 'ManyToMany' | 'ManyToOne';
      // 字段映射
      mapper: Map<string, string>;
    }
  > = new Map();

  for (const property of properties) {
    if (property.type.type !== DataObjectTypeName.Reference) {
      continue;
    }

    const { target, targetProperty, cardinality } = property.type as Required<ViewDSL.DataObjectReference>;

    // 自我引用去掉?
    if (target === viewDSL.uuid) {
      continue;
    }

    const targetObject = getObject(target);
    const targetPropertyInfo = getReference(target, targetProperty);
    const targetName = targetObject.name;

    let state = targets.get(targetName);

    if (state == null) {
      state = {
        cardinality,
        mapper: new Map(),
      };
      targets.set(targetName, state);
    }

    state.cardinality ??= cardinality;
    state.mapper.set(property.name, targetPropertyInfo.name);
  }

  return {
    source: viewDSL.name,
    targets: Array.from(targets.entries()).map(([target, { cardinality, mapper }]) => ({
      target,
      cardinality,
      mapper: Array.from(mapper.entries()).map(([sourceField, targetField]) => ({
        sourceField,
        targetField,
      })),
    })),
  };
}

export class DataObject extends Node<ViewDSL.DataObjectDSL> {
  toDSL(): DSL.DataObjectDSL {
    const { uuid, title, name, description, meta, properties, tableName, indexes } = this.properties;

    const transformedIndexes = indexes.map(i => transformIndex(i, this.getProperty));

    // 收集主键索引
    const primaryKeys = this.properties.properties.filter(i => i.primaryKey);
    if (primaryKeys.length) {
      transformedIndexes.push({
        uuid: `${uuid}-primary-key`,
        name: 'primaryKeys',
        type: 'Primary',
        properties: primaryKeys.map(i => i.name),
      });
    }

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      tableName: transformString(tableName),
      properties: properties.map(i => transformProperty(i, this.getPropertyReference)),
      indexes: transformedIndexes,
    };
  }

  toReferenceDSL(): DSL.DataObjectReferenceDSL {
    return transformDataObjectReference(this.properties, this.getObject, this.getPropertyReference);
  }

  getObject = (id: string) => {
    return this.getReference(id) as ViewDSL.DataObjectDSL;
  };

  getProperty = (id: string): ViewDSL.DataObjectPropertyDSL => {
    return this.properties.properties.find(i => i.uuid === id)!;
  };

  getPropertyReference = (id: string, propertyId: string) => {
    const node = this.container.getNodeById(id) as unknown as DataObject;

    return node.properties.properties.find(i => i.uuid === propertyId)!;
  };
}

/**
 * 数据对象容器
 */
export class DataModelContainer extends BaseContainer implements IContainer {
  dataObjects: DataObject[] = [];

  /**
   * 收集所有有效的节点
   */
  nodesIndexByUUID: Map<string, DataObject> = new Map();

  constructor(tree: Record<string, Tree>) {
    super();

    this.traverse(tree);
  }

  toDSL(): DSL.DataModelDSL {
    return {
      dataObjects: this.dataObjects.map(i => i.toDSL()),
      references: this.dataObjects.map(i => i.toReferenceDSL()).filter(i => i.targets.length),
    };
  }

  getNodeById(id: string): Node<any> | undefined {
    return this.nodesIndexByUUID.get(id);
  }

  handle(node: Tree, tree: Record<string, Tree>): void {
    const type = node.properties.__node_name__;

    if (type !== DataObjectName.DataObject) {
      return;
    }

    const properties = node.properties as unknown as ViewDSL.DataObjectDSL;
    const object = new DataObject(properties.uuid, properties, this);
    this.dataObjects.push(object);
    this.nodesIndexByUUID.set(object.id, object);
  }
}
