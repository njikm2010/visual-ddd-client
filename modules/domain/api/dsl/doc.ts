import { Doc as YDoc, applyUpdate } from 'yjs';

import { YJS_FIELD_NAME } from '../../constants';

import * as DSL from './interface';
import { transform as transformToDomainDSL } from './domain-model';
import { transform as transformToQueryDSL } from './query-model';
import { transform as transformToDataObjectDSL } from './data-model';

/**
 * 创建文档
 * @param update
 * @returns
 */
export function createDoc(update: Uint8Array) {
  const doc = new YDoc();

  applyUpdate(doc, update);

  return doc;
}

/**
 * 转换为 DSL
 * @param doc
 * @returns
 */
export function transformToDSL(doc: YDoc): DSL.BusinessDomainDSL {
  const domainMap = doc.getMap(YJS_FIELD_NAME.DOMAIN);
  const queryMap = doc.getMap(YJS_FIELD_NAME.QUERY);
  const dataObjectMap = doc.getMap(YJS_FIELD_NAME.DATA_OBJECT);

  return {
    domainModel: transformToDomainDSL(domainMap.toJSON()),
    queryModel: transformToQueryDSL(queryMap.toJSON()),
    dataModel: transformToDataObjectDSL(dataObjectMap.toJSON()),
  };
}
