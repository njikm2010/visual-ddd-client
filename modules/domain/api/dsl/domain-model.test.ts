import fs from 'fs';
import path from 'path';
import * as ViewDSL from '@/modules/domain/domain-design/dsl/dsl';
import {
  DomainModelContainer,
  transformMeta,
  transformType,
  transformProperty,
  transformMethods,
  transformEnum,
  transformRule,
  transformParameter,
} from './domain-model';
import { Tree, BaseContainer } from './shared';

test('transformMeta', () => {
  expect(transformMeta([])).toEqual({});
  expect(transformMeta([{ key: '', value: '' }])).toEqual({});
  expect(transformMeta([{ key: 'ok', value: '' }])).toEqual({});
  expect(transformMeta([{ key: 'ok', value: 'yes' }])).toEqual({ ok: 'yes' });
});

test('transformType', () => {
  const getRef = (id: string) => ({ name: `NameOf${id}`, uuid: id });
  expect(transformType({ type: ViewDSL.TypeType.Base, name: 'Boolean' }, getRef)).toBe('Boolean');
  expect(
    transformType(
      {
        type: ViewDSL.TypeType.Container,
        name: 'List',
        params: { item: { type: ViewDSL.TypeType.Base, name: 'Boolean' } },
      },
      getRef
    )
  ).toBe('List<Boolean>');

  expect(
    transformType(
      {
        type: ViewDSL.TypeType.Container,
        name: 'Map',
        params: {
          key: { type: ViewDSL.TypeType.Base, name: 'Boolean' },
          value: {
            type: ViewDSL.TypeType.Reference,
            referenceId: 'Foo',
            name: 'Foo',
          },
        },
      },
      getRef
    )
  ).toBe('Map<Boolean, [NameOfFoo:Foo]>');

  expect(
    transformType(
      {
        type: ViewDSL.TypeType.Container,
        name: 'Map',
        params: {
          key: { type: ViewDSL.TypeType.Base, name: 'Boolean' },
          value: {
            type: ViewDSL.TypeType.Container,
            name: 'List',
            params: {
              item: {
                type: ViewDSL.TypeType.Reference,
                referenceId: 'Foo',
                name: 'Foo',
              },
            },
          },
        },
      },
      getRef
    )
  ).toBe('Map<Boolean, List<[NameOfFoo:Foo]>>');
});

test('transformProperty', () => {
  const getRef = (id: string) => ({ name: `NameOf${id}`, uuid: id });
  expect(
    transformProperty(
      {
        uuid: '1',
        title: 'ok',
        name: 'ok',
        description: 'ok',
        access: 'protected',
        optional: true,
        type: { type: ViewDSL.TypeType.Base, name: 'Boolean' },
      },
      getRef
    )
  ).toEqual({
    access: 'protected',
    description: 'ok',
    meta: undefined,
    name: 'ok',
    title: 'ok',
    type: 'Boolean',
    optional: true,
    uuid: '1',
  });

  expect(
    transformProperty(
      {
        uuid: '1',
        title: 'ok',
        name: 'ok',
        description: 'ok',
      },
      getRef
    )
  ).toEqual({
    description: 'ok',
    meta: undefined,
    name: 'ok',
    title: 'ok',
    type: 'Void',
    uuid: '1',
  });
});

test('transformParameter', () => {
  const getRef = (id: string) => ({ name: `NameOf${id}`, uuid: id });
  expect(
    transformParameter(
      {
        uuid: '1',
        title: 'ok',
        name: 'ok',
        description: 'ok',
        optional: true,
        type: { type: ViewDSL.TypeType.Base, name: 'Boolean' },
      },
      getRef
    )
  ).toEqual({
    description: 'ok',
    meta: undefined,
    name: 'ok',
    title: 'ok',
    type: 'Boolean',
    optional: true,
    uuid: '1',
  });

  expect(
    transformParameter(
      {
        uuid: '1',
        title: 'ok',
        name: 'ok',
        description: 'ok',
      },
      getRef
    )
  ).toEqual({
    description: 'ok',
    meta: undefined,
    name: 'ok',
    title: 'ok',
    type: 'Void',
    uuid: '1',
  });
});

test('transformMethods', () => {
  const getRef = () => ({ name: 'Foo', uuid: 'ok' });

  expect(transformMethods([], getRef)).toEqual([]);
  expect(
    transformMethods(
      [
        {
          uuid: '0',
          name: 'foo',
          parameters: [],
        },
      ],
      getRef
    )
  ).toEqual([
    {
      abstract: undefined,
      access: undefined,
      description: undefined,
      meta: undefined,
      name: 'foo',
      signature: {
        uuid: '0',
        description: undefined,
        parameters: [],
        return: {
          type: 'void',
        },
      },
      title: undefined,
      uuid: '0-method',
    },
  ]);

  expect(() => {
    transformMethods(
      [
        {
          uuid: '0',
          name: 'foo',
          parameters: [],
        },
        {
          uuid: '2',
          name: 'foo',
          parameters: [{ uuid: '2-1', name: 'a', type: { type: ViewDSL.TypeType.Base, name: 'Boolean' } }],
          result: { type: ViewDSL.TypeType.Base, name: 'Char' },
        },
      ],
      getRef
    );
  }).toThrowError();

  //   .toEqual([
  //   {
  //     abstract: undefined,
  //     access: undefined,
  //     description: undefined,
  //     meta: undefined,
  //     name: 'foo',
  //     signature: [
  //       {
  //         description: undefined,
  //         parameters: [],
  //         return: {
  //           type: 'void',
  //         },
  //       },
  //       {
  //         description: undefined,
  //         parameters: [
  //           {
  //             description: undefined,
  //             meta: undefined,
  //             name: 'a',
  //             title: undefined,
  //             type: 'Boolean',
  //             uuid: '2-1',
  //           },
  //         ],
  //         return: {
  //           type: 'Char',
  //         },
  //       },
  //     ],
  //     title: undefined,
  //     uuid: '0',
  //   },
  // ]);
});

test('transformEnum', () => {
  expect(
    transformEnum({ uuid: '1', name: 'ok', title: 'ok', description: 'ok', baseType: 'number', members: [] })
  ).toEqual({
    baseType: 'number',
    description: 'ok',
    members: [],
    meta: undefined,
    name: 'ok',
    title: 'ok',
    uuid: '1',
  });

  expect(
    transformEnum({
      uuid: '1',
      name: 'ok',
      title: 'ok',
      description: 'ok',
      baseType: 'number',
      members: [
        {
          uuid: '1-1',
          name: 'Ok',
          code: '1',
        },
      ],
    })
  ).toEqual({
    baseType: 'number',
    description: 'ok',
    members: [
      {
        code: 1,
        description: undefined,
        meta: undefined,
        name: 'Ok',
        title: undefined,
        uuid: '1-1',
      },
    ],
    meta: undefined,
    name: 'ok',
    title: 'ok',
    uuid: '1',
  });
});

test('transformRule', () => {
  expect(
    transformRule({
      uuid: '1',
      title: 'ok',
      name: 'ok',
      description: 'desc',
    })
  ).toEqual({
    description: 'desc',
    meta: undefined,
    name: 'ok',
    title: 'ok',
    uuid: '1',
  });
});

const DATA = {
  __ROOT__: {
    __NODE__: true,
    id: '__ROOT__',
    children: {
      '3cdd4fd4-9f49-4747-b85d-2898759a880d': 1,
      '5a2fe372-faa6-42ce-912e-992809a358e9': 1,
      'copy-0026add6-8a2a-4e26-bc08-2c5d95443445': 1,
      'f7905b48-9e01-407c-a41b-8a8bebfab920': 1,
      'copy-44a87083-8392-499e-8c01-a7adbdaa9a34': 1,
      'copy-78a9324b-773f-4cd4-9690-db995602e492': 1,
    },
    properties: { __node_name__: '__ROOT__', __node_type__: 'node', __PROPERTY__: true },
  },
  '3cdd4fd4-9f49-4747-b85d-2898759a880d': {
    __NODE__: true,
    id: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
    parent: '__ROOT__',
    children: {
      '89772ddd-2140-4ef6-b330-52d465f8cc78': 1,
      'copy-db705261-7462-4c77-88e3-ac9c782534d1': 1,
      '499bfd1e-e43b-4310-af32-c5d95b3f97d6': 1,
      '918dddbf-a695-4760-b4ad-8dfd1b66571f': 1,
    },
    properties: {
      __node_name__: 'aggregation',
      __node_type__: 'node',
      position: { x: 91, y: 420 },
      uuid: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      color: '#D9F7BE',
      zIndex: 1,
      size: { width: 586, height: 391.125 },
      __prevent_auto_resize__: true,
      originSize: { width: 500, height: 300 },
      __PROPERTY__: true,
    },
  },
  '89772ddd-2140-4ef6-b330-52d465f8cc78': {
    __NODE__: true,
    id: '89772ddd-2140-4ef6-b330-52d465f8cc78',
    parent: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
    children: {},
    properties: {
      __node_name__: 'value-object',
      __node_type__: 'node',
      position: { x: 152, y: 675 },
      uuid: '89772ddd-2140-4ef6-b330-52d465f8cc78',
      name: 'ValueObject',
      title: '未命名',
      description: '',
      meta: [],
      implements: [],
      abstract: false,
      properties: [
        {
          uuid: 'e6b0f980-ca73-4af9-956e-bec51d1b66cc',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      methods: [],
      classProperties: [],
      classMethods: [],
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  '5a2fe372-faa6-42ce-912e-992809a358e9': {
    __NODE__: true,
    id: '5a2fe372-faa6-42ce-912e-992809a358e9',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'command',
      __node_type__: 'node',
      position: { x: 270, y: -15 },
      uuid: '5a2fe372-faa6-42ce-912e-992809a358e9',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      source: {
        http: { enabled: true },
        rpc: { enabled: true },
        event: { enabled: false, value: '' },
        schedule: { enabled: false, value: '' },
      },
      properties: [
        {
          uuid: 'dfe5dff7-2035-4363-9839-29ff4fd103cc',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      eventProperties: [],
      repository: 'modify',
      eventSendable: false,
      aggregation: { referenceId: '3cdd4fd4-9f49-4747-b85d-2898759a880d', name: '未命名(Untitled)' },
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  'copy-0026add6-8a2a-4e26-bc08-2c5d95443445': {
    __NODE__: true,
    id: 'copy-0026add6-8a2a-4e26-bc08-2c5d95443445',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'command',
      __node_type__: 'node',
      uuid: 'copy-0026add6-8a2a-4e26-bc08-2c5d95443445',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      source: {
        http: { enabled: true },
        rpc: { enabled: true },
        event: { enabled: false, value: '' },
        schedule: { enabled: false, value: '' },
      },
      properties: [
        {
          uuid: 'dfe5dff7-2035-4363-9839-29ff4fd103cc',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      eventProperties: [],
      repository: 'modify',
      eventSendable: false,
      zIndex: 2,
      position: { x: 555, y: 0 },
      size: { width: 200, height: 129.96875 },
      __PROPERTY__: true,
    },
  },
  'f7905b48-9e01-407c-a41b-8a8bebfab920': {
    __NODE__: true,
    id: 'f7905b48-9e01-407c-a41b-8a8bebfab920',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'rule',
      __node_type__: 'node',
      position: { x: 210, y: 204 },
      uuid: 'f7905b48-9e01-407c-a41b-8a8bebfab920',
      name: 'Untitled1',
      title: '未命名',
      description: '规则描述',
      meta: [],
      aggregator: { referenceId: '5a2fe372-faa6-42ce-912e-992809a358e9', name: '未命名(Untitled)' },
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  'copy-44a87083-8392-499e-8c01-a7adbdaa9a34': {
    __NODE__: true,
    id: 'copy-44a87083-8392-499e-8c01-a7adbdaa9a34',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'rule',
      __node_type__: 'node',
      uuid: 'copy-44a87083-8392-499e-8c01-a7adbdaa9a34',
      name: 'Untitled2',
      title: '未命名',
      description: '规则描述',
      meta: [],
      aggregator: { referenceId: '5a2fe372-faa6-42ce-912e-992809a358e9', name: '未命名(Untitled)' },
      zIndex: 2,
      position: { x: 384, y: 204 },
      size: { width: 150, height: 94.125 },
      __PROPERTY__: true,
    },
  },
  'copy-78a9324b-773f-4cd4-9690-db995602e492': {
    __NODE__: true,
    id: 'copy-78a9324b-773f-4cd4-9690-db995602e492',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'rule',
      __node_type__: 'node',
      uuid: 'copy-78a9324b-773f-4cd4-9690-db995602e492',
      name: 'Untitled2',
      title: '未命名',
      description: '规则描述',
      meta: [],
      aggregator: { referenceId: 'copy-0026add6-8a2a-4e26-bc08-2c5d95443445', name: '未命名(Untitled)' },
      zIndex: 2,
      position: { x: 614, y: 204 },
      size: { width: 150, height: 94.125 },
      __PROPERTY__: true,
    },
  },
  'copy-db705261-7462-4c77-88e3-ac9c782534d1': {
    __NODE__: true,
    id: 'copy-db705261-7462-4c77-88e3-ac9c782534d1',
    parent: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
    children: {},
    properties: {
      __node_name__: 'entity',
      __node_type__: 'node',
      uuid: 'copy-db705261-7462-4c77-88e3-ac9c782534d1',
      name: 'Untitled2',
      title: '未命名',
      description: '',
      meta: [],
      implements: [],
      abstract: false,
      properties: [
        {
          uuid: 'c1df59f2-5d44-41fd-80ff-fabfaab1d255',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'reference', referenceId: '918dddbf-a695-4760-b4ad-8dfd1b66571f', name: 'MyEnum' },
          access: 'public',
        },
      ],
      methods: [
        {
          uuid: '53afcda5-ad8e-45a0-887b-1a23620f2235',
          name: 'untitled',
          title: '',
          description: '',
          meta: [],
          access: 'public',
          abstract: false,
          parameters: [],
        },
      ],
      classProperties: [],
      classMethods: [],
      isAggregationRoot: false,
      id: 'c1df59f2-5d44-41fd-80ff-fabfaab1d255',
      zIndex: 2,
      position: { x: 369, y: 503 },
      size: { width: 200, height: 118.96875 },
      __PROPERTY__: true,
    },
  },
  '499bfd1e-e43b-4310-af32-c5d95b3f97d6': {
    __NODE__: true,
    id: '499bfd1e-e43b-4310-af32-c5d95b3f97d6',
    parent: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
    children: {},
    properties: {
      __node_name__: 'entity',
      __node_type__: 'node',
      position: { x: 137, y: 503 },
      uuid: '499bfd1e-e43b-4310-af32-c5d95b3f97d6',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      implements: [],
      abstract: false,
      properties: [
        {
          uuid: '0c9ad68c-19de-459e-be69-57d2c07c6dfa',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      methods: [
        {
          uuid: 'f1559572-9a47-43a7-8df7-c2a690c1e57b',
          name: 'untitled',
          title: '',
          description: '',
          meta: [],
          access: 'public',
          abstract: false,
          parameters: [],
        },
      ],
      classProperties: [],
      classMethods: [],
      isAggregationRoot: true,
      id: '0c9ad68c-19de-459e-be69-57d2c07c6dfa',
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  '918dddbf-a695-4760-b4ad-8dfd1b66571f': {
    __NODE__: true,
    id: '918dddbf-a695-4760-b4ad-8dfd1b66571f',
    parent: '3cdd4fd4-9f49-4747-b85d-2898759a880d',
    children: {},
    properties: {
      __node_name__: 'enum',
      __node_type__: 'node',
      position: { x: 432, y: 675 },
      uuid: '918dddbf-a695-4760-b4ad-8dfd1b66571f',
      name: 'MyEnum',
      title: '未命名',
      description: '',
      meta: [],
      baseType: 'number',
      members: [
        {
          uuid: '39f48e70-09bf-4f53-93b4-04a550a6e286',
          name: 'UNTITLED',
          title: '',
          description: '',
          meta: [],
          code: '0',
        },
      ],
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
};

test('BaseContainer', () => {
  class MyContainer extends BaseContainer {
    ids: string[] = [];

    handle(node: Tree<any>): void {
      this.ids.push(node.id);
    }
  }

  const c = new MyContainer();
  c.traverse(DATA as any);

  expect(c.ids).toMatchSnapshot();
});

test('Container', () => {
  const container = new DomainModelContainer(DATA as any);

  expect(container.toDSL()).toMatchSnapshot();
});

test('real data', async () => {
  const data = JSON.parse((await fs.promises.readFile(path.resolve(__dirname, './domain-model.data.json'))).toString());

  const container = new DomainModelContainer(data as any);

  expect(container.toDSL()).toMatchSnapshot();
});
