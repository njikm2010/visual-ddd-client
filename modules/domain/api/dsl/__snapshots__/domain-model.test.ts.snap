// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseContainer 1`] = `
[
  "__ROOT__",
  "3cdd4fd4-9f49-4747-b85d-2898759a880d",
  "89772ddd-2140-4ef6-b330-52d465f8cc78",
  "copy-db705261-7462-4c77-88e3-ac9c782534d1",
  "499bfd1e-e43b-4310-af32-c5d95b3f97d6",
  "918dddbf-a695-4760-b4ad-8dfd1b66571f",
  "5a2fe372-faa6-42ce-912e-992809a358e9",
  "copy-0026add6-8a2a-4e26-bc08-2c5d95443445",
  "f7905b48-9e01-407c-a41b-8a8bebfab920",
  "copy-44a87083-8392-499e-8c01-a7adbdaa9a34",
  "copy-78a9324b-773f-4cd4-9690-db995602e492",
]
`;

exports[`Container 1`] = `
{
  "aggregates": [
    {
      "commands": [
        {
          "category": undefined,
          "description": "",
          "event": {
            "name": "UntitledEvent",
            "properties": [],
            "uuid": "5a2fe372-faa6-42ce-912e-992809a358e9",
          },
          "eventSendable": false,
          "meta": {},
          "name": "Untitled",
          "properties": [
            {
              "access": "public",
              "description": "",
              "meta": {},
              "name": "untitled",
              "title": "未命名",
              "type": "String",
              "uuid": "dfe5dff7-2035-4363-9839-29ff4fd103cc",
            },
          ],
          "repository": "modify",
          "return": {
            "type": "void",
          },
          "rules": [
            {
              "description": "规则描述",
              "meta": {},
              "name": "Untitled1",
              "title": "未命名",
              "uuid": "f7905b48-9e01-407c-a41b-8a8bebfab920",
            },
            {
              "description": "规则描述",
              "meta": {},
              "name": "Untitled2",
              "title": "未命名",
              "uuid": "copy-44a87083-8392-499e-8c01-a7adbdaa9a34",
            },
          ],
          "source": [
            {
              "type": "http",
            },
            {
              "type": "rpc",
            },
          ],
          "title": "未命名",
          "uuid": "5a2fe372-faa6-42ce-912e-992809a358e9",
        },
      ],
      "description": "",
      "entities": [
        {
          "description": "",
          "id": {
            "access": "public",
            "description": "",
            "meta": {},
            "name": "untitled",
            "title": "未命名",
            "type": "[MyEnum:918dddbf-a695-4760-b4ad-8dfd1b66571f]",
            "uuid": "c1df59f2-5d44-41fd-80ff-fabfaab1d255",
          },
          "meta": {},
          "methods": [
            {
              "abstract": false,
              "access": "public",
              "description": "",
              "meta": {},
              "name": "untitled",
              "signature": {
                "description": "",
                "parameters": [],
                "return": {
                  "type": "void",
                },
              },
              "title": "",
              "uuid": "53afcda5-ad8e-45a0-887b-1a23620f2235",
            },
          ],
          "name": "Untitled2",
          "properties": [],
          "title": "未命名",
          "uuid": "copy-db705261-7462-4c77-88e3-ac9c782534d1",
        },
      ],
      "enums": [
        {
          "baseType": "number",
          "description": "",
          "members": [
            {
              "code": 0,
              "description": "",
              "meta": {},
              "name": "UNTITLED",
              "title": "",
              "uuid": "39f48e70-09bf-4f53-93b4-04a550a6e286",
            },
          ],
          "meta": {},
          "name": "MyEnum",
          "title": "未命名",
          "uuid": "918dddbf-a695-4760-b4ad-8dfd1b66571f",
        },
      ],
      "meta": {},
      "name": "Untitled",
      "root": {
        "description": "",
        "id": {
          "access": "public",
          "description": "",
          "meta": {},
          "name": "untitled",
          "title": "未命名",
          "type": "String",
          "uuid": "0c9ad68c-19de-459e-be69-57d2c07c6dfa",
        },
        "meta": {},
        "methods": [
          {
            "abstract": false,
            "access": "public",
            "description": "",
            "meta": {},
            "name": "untitled",
            "signature": {
              "description": "",
              "parameters": [],
              "return": {
                "type": "void",
              },
            },
            "title": "",
            "uuid": "f1559572-9a47-43a7-8df7-c2a690c1e57b",
          },
        ],
        "name": "Untitled",
        "properties": [],
        "title": "未命名",
        "uuid": "499bfd1e-e43b-4310-af32-c5d95b3f97d6",
      },
      "title": "未命名",
      "uuid": "3cdd4fd4-9f49-4747-b85d-2898759a880d",
      "valueObjects": [
        {
          "description": "",
          "meta": {},
          "methods": [],
          "name": "ValueObject",
          "properties": [
            {
              "access": "public",
              "description": "",
              "meta": {},
              "name": "untitled",
              "title": "未命名",
              "type": "String",
              "uuid": "e6b0f980-ca73-4af9-956e-bec51d1b66cc",
            },
          ],
          "title": "未命名",
          "uuid": "89772ddd-2140-4ef6-b330-52d465f8cc78",
        },
      ],
    },
  ],
}
`;
